### PR TITLE
Enhance From File catalog loading to support more columns and improve Clear Table functionality

### PR DESCRIPTION
**Pull request by haticekaratay**
### Description

This pull request introduces:
   Catalog Loading from File:
      - Users can now load catalog data (e.g., .ecsv files) directly into the table.
   Improved Clear Table Functionality:
      - The "Clear Table" button now clears the table, removes markers, and resets selected rows in the viewer.

https://github.com/user-attachments/assets/691d56f2-9d89-4dea-8987-c517733564fd

Fixes #<Issue Number>

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/core/template_mixin.py [hunk 1]
```diff
@@ -4691,12 +4691,20 @@ def float_precision(column, item):
...
                 return ''
-
             elif isinstance(item, float):
                 return float_precision(column, item)
             elif isinstance(item, (list, tuple)):
                 return [float_precision(column, i) if isinstance(i, float) else i for i in item]
-
+            elif isinstance(item, (np.float32, np.float64)):
+                return float(item)
+            elif isinstance(item, u.Quantity):
```
Comment (pllim) [inline jdaviz/core/template_mixin.py:4700]: This way we lost the unit info, is that important?

Also I wonder if we can take advantage of the JsonCsutomEncoder that astropy already has for some of these: https://github.com/astropy/astropy/blob/93b6f551b9ed00c4a422b754eae0261f1a7552c9/astropy/utils/misc.py#L366

File: jdaviz/configs/imviz/tests/test_catalogs.py [hunk 2]
```diff
@@ -157,12 +158,16 @@ def test_plugin_image_with_result(self, imviz_helper, tmp_path):
...
+        catalogs_plugin.table.selected_rows = [
+            catalogs_plugin.table.items[0]]
+        # Then zoom in to the selected
         catalogs_plugin.vue_zoom_in()
 
-        assert imviz_helper.viewers['imviz-0']._obj.state.x_min == 858.24969
-        assert imviz_helper.viewers['imviz-0']._obj.state.x_max == 958.38461
-        assert imviz_helper.viewers['imviz-0']._obj.state.y_min == 278.86265
-        assert imviz_helper.viewers['imviz-0']._obj.state.y_max == 378.8691
+        assert imviz_helper.viewers['imviz-0']._obj.state.x_min == 1022.5631800000001
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_catalogs.py:167]: Why did the test results change?

File: CHANGES.rst [hunk 3]
```diff
@@ -21,6 +21,8 @@ Imviz
 - Catalog Search plugin now exposes a maximum sources limit for all catalogs and resolves an edge case
   when loading a catalog from a file that only contains one source. [#3337]
 
+- Catalog Search plugin now supports loading catalog data directly from files. [#3359]
```
Comment (kecnry) [inline CHANGES.rst:24]: can we clarify somehow that this adds support for the additional columns (otherwise I worry we might get confused that this is when support for loading from file was first introduced at all)?

**Comment by pllim** [comment]
We should also probably add a section in user docs to document exactly what format of such file the plugin can actually ingest?

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 4]
```diff
@@ -88,13 +89,15 @@ def __init__(self, *args, **kwargs):
     def _file_parser(path):
```
Comment (gibsongreen) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:89]: The docs say that the `label` column must exist, but we don't actually enforce that so you are able to load with just the `sky_centroid` column. This results in the `Object ID` column = N/A, so I think we should add a condition here so we make sure that column also exists.

Comment (haticekaratay) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:91]: Thank you for your comment! Since `sky_coord` is sufficient to display the data on the viewer, I have updated the documentation to reflect that the label column is now optional. Additionally, I implemented logic to assign default values to the `Object ID` column when the `label` column is not present.

**Review by pllim — COMMENTED** [review]
Some technical comments/questions.

I like the new test you added that uses local data. Thanks!

**Comment by pllim** [comment]
Somehow the changes here cause ResourceWarning in a catalog test. Unclosed file somewhere? Should investigate.

```
FAILED jdaviz/configs/imviz/tests/test_catalogs.py::TestCatalogs::test_plugin_image_with_result
...
ResourceWarning: unclosed <ssl.SSLSocket ...>
```

**Comment by cshanahan1** [comment]
If you write out a table from the app and try to read it back in, it complains that theres not a 'sky centroid' column even though it has an RA and dec

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 5]
```diff
@@ -271,16 +279,19 @@ def search(self, error_on_fail=False):
             if len(self.app._catalog_source_table) == 1 or self.max_sources == 1:
                 x_coordinates = [x_coordinates]
                 y_coordinates = [y_coordinates]
+            for idx, (row, x_coord, y_coord) in enumerate(zip(self.app._catalog_source_table, x_coordinates, y_coordinates)):  # noqa:E501
+                row_info = {
```
Comment (cshanahan1) [inline jdaviz/configs/imviz/plugins/catalogs/catalogs.py:283]: the loaded table has to have a 'sky centroid' column, but it is then broken up into ra and dec, which means when its written back out it cant be read back in because it doesn't have 'sky centroid'. would it be possible to add a check when loading if the table already contains a ra and dec column rather than unpacking sky centroid? as it is, a table written out by the app can not be read back in

Comment (pllim) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:283]: Does that problem pre-date this PR? Just wondering if round-tripping is in scope here (e.g., writing it out back as sky centroid as a single column in ECSV format).

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 6]
```diff
@@ -204,6 +207,11 @@ def search(self, error_on_fail=False):
             # all exceptions when going through the UI should have prevented setting this path
             # but this exceptions might be raised here if setting from_file from the UI
             table = self.catalog.selected_obj
+            column_names = list(table.colnames)
+            self.table.headers_avail = self.headers + [
+                col for col in column_names if col not in self.headers and col not in ["sky_centroid", "label"]     # noqa:E501
```
[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:212]: p.p.p.s. Please ignore the p.p.s. (I crossed it out) because `set` would lose the insertion order. Thanks.

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py (hunk 5 above)
Comment (gibsongreen) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:283]: `export_table` is ultimately called which will treat the the `TableMixin` columns as a `QTable` https://github.com/spacetelescope/jdaviz/blob/48be1548a8b00be113966fbc8c354dd6fdac1026/jdaviz/core/template_mixin.py#L4790 and write the columns as is, so unless you have a column in our table that is labeled `sky_centroid`, you wouldn't be able to read it back in after exporting. This is also true if you try exporting Gaia or SDSS.

In the case of 'from file', we already have the `sky_centroid` column, so we could add the sky_centroid column with the same name to our table, export the table, and then the exported file is able to be reloaded into Imviz.

The other catalogs options could be added in a follow up effort for either writing the `sky_centroid` column to the loaded catalog table or handle it on export.

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 7]
```diff
@@ -90,13 +91,15 @@ def __init__(self, *args, **kwargs):
...
         try:
             table = QTable.read(path)
-        except Exception:
-            return 'Could not parse file with astropy.table.QTable.read', {}
+            if not table.colnames:  # Ensure the file has columns
+                return "File contains no columns", {}
 
-        if 'sky_centroid' not in table.colnames:
-            return 'Table does not contain required sky_centroid column', {}
+            if 'sky_centroid' not in table.colnames:
```
[the remark this answers is not in the document]
Comment (haticekaratay) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py]: Thank you for catching that! Honestly, I don’t know what I was thinking when I moved those checks inside the try-except block. You’re absolutely right that they belong outside since they’re not handling exceptions from QTable.read. I’ve fixed it now, and I appreciate your sharp eye for detail!

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py (hunk 5 above)
Comment (haticekaratay) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:283]: I was trying to keep it aligned with the other `Search` options, but I didn't think of the export functionality then. Thanks for the additional context, all; I will keep the existing `sky_centroid` column in the table.

File: jdaviz/configs/imviz/plugins/catalogs/catalogs.py [hunk 8]
```diff
@@ -402,9 +412,13 @@ def vue_do_search(self, *args, **kwargs):
         # calls self.search() which handles all of the searching logic
         self.search()
 
-    def clear(self, hide_only=True):
+    def clear_table(self, hide_only=True):
```
[the remark this answers is not in the document]
Comment (haticekaratay) [reply jdaviz/configs/imviz/plugins/catalogs/catalogs.py:418]: Thanks for the catch. Yes, this is an explicit overwrite. I updated the change log.

File: jdaviz/configs/imviz/tests/test_catalogs.py [hunk 9]
```diff
@@ -158,14 +158,25 @@ def test_plugin_image_with_result(self, imviz_helper, tmp_path):
...
         catalogs_plugin.zoom_to_selected(padding=50 / 2048)
 
-        assert imviz_helper.viewers['imviz-0']._obj.state.x_min == 858.24969
-        assert imviz_helper.viewers['imviz-0']._obj.state.x_max == 958.38461
-        assert imviz_helper.viewers['imviz-0']._obj.state.y_min == 278.86265
-        assert imviz_helper.viewers['imviz-0']._obj.state.y_max == 378.8691
+        assert_allclose(
+            imviz_helper.viewers['imviz-0']._obj.state.x_min, 1022.57570000, atol=0.1)
+        assert_allclose(imviz_helper.viewers['imviz-0']._obj.state.x_max,
+                        1122.5757, atol=0.1)
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_catalogs.py:175]: Consistent styling with the other 3 checks to make it easier to read.
```suggestion
        assert_allclose(
            imviz_helper.viewers['imviz-0']._obj.state.x_max, 1122.5757, atol=0.1)
```

File: jdaviz/configs/imviz/tests/test_catalogs.py [hunk 10]
```diff
@@ -240,27 +254,34 @@ def test_offline_ecsv_catalog(imviz_helper, image_2d_wcs, tmp_path):
...
     # test the zooming using the default 'padding' of 2% of the viewer size
     # around selected points
     catalogs_plugin.zoom_to_selected()
-    assert imviz_helper.viewers['imviz-0']._obj.state.x_min == -0.19966
-    assert imviz_helper.viewers['imviz-0']._obj.state.x_max == 0.20034000000000002
-    assert imviz_helper.viewers['imviz-0']._obj.state.y_min == 0.8000100000000001
-    assert imviz_helper.viewers['imviz-0']._obj.state.y_max == 1.20001
+    assert_allclose(imviz_helper.viewers['imviz-0']._obj.state.x_min, -0.19966, rtol=1e-1)
+    assert_allclose(imviz_helper.viewers['imviz-0']._obj.state.x_max,
+                    0.20034000000000002, rtol=1e-1)
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_catalogs.py:282]: Why did this go from exact comparison to really loose allclose with 10% tolerance?

```suggestion
    assert_allclose(imviz_helper.viewers['imviz-0']._obj.state.x_max, 0.20034000000000002, rtol=1e-1)
```

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

**Review by gibsongreen — APPROVED** [review]
Everything looks good to me, great work!

**Comment by pllim** [comment]
@meeseeksdev backport to v4.1.x
